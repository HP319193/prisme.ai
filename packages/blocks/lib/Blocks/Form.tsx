import { Button, SchemaForm } from '@prisme.ai/design-system';
import { useBlock } from '../Provider';
import { useTranslation } from 'react-i18next';
import { useCallback, useMemo } from 'react';
import { tw } from 'twind';
import BlockTitle from './Internal/BlockTitle';
import useLocalizedTextBlock from '../useLocalizedTextBlock';

const defaultSchema = {
  type: 'string',
  title: 'preview',
};

export const Form = () => {
  const { config = {}, events } = useBlock();
  const { t } = useTranslation('pages');
  const { localize, localizeSchemaForm } = useLocalizedTextBlock();

  const onChange = useCallback(
    (values: any) => {
      if (!config.onChange || !events) return;
      events.emit(config.onChange, values);
    },
    [config.onChange, events]
  );

  const onSubmit = useCallback(
    (values: any) => {
      if (!config.onSubmit || !events) return;
      events.emit(config.onSubmit, values);
    },
    [config.onSubmit, events]
  );

  const localizedSchema = useMemo(() => {
    return localizeSchemaForm(config.schema || defaultSchema);
  }, [config.schema, localizeSchemaForm]);

  return (
    <div className={tw`p-8`}>
      {config.title && <BlockTitle value={config.title} />}
      <SchemaForm
        schema={localizedSchema}
        onChange={onChange}
        onSubmit={onSubmit}
        buttons={[
          <div key={0} className={tw`flex grow justify-end mt-2 pt-4`}>
            <Button
              type="submit"
              variant="primary"
              className={tw`!py-4 !px-8 !h-max`}
            >
              {localize(config.submitLabel) || t('blocks.form.submit')}
            </Button>
          </div>,
        ]}
      />
    </div>
  );
};

export default Form;
