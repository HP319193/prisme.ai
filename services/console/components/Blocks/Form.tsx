import {
  Button,
  Schema,
  SchemaForm,
  Tooltip,
  useBlock,
} from '@prisme.ai/design-system';
import { FieldProps } from '@prisme.ai/design-system/lib/Components/SchemaForm/types';
import { useTranslation } from 'next-i18next';
import { useCallback, useMemo } from 'react';
import { useField } from 'react-final-form';
import useLocalizedText from '../../utils/useLocalizedText';
import BlockTitle from './Components/BlockTitle';
import { InfoCircleOutlined } from '@ant-design/icons';
import Properties from '../SchemaFormBuilder/Properties';

const defaultSchema = {
  type: 'string',
  title: 'preview',
};

export const SchemaEditor = ({ name }: FieldProps) => {
  const field = useField(name);
  const { t } = useTranslation('workspaces');

  return (
    <div className="flex flex-1 mt-4">
      <div className="ant-input ">
        <div className="text-gray pl-4 flex flex-1 flex-row justify-between">
          {t('pages.blocks.form.schema.label')}
          <div className="text-accent mr-2">
            <Tooltip
              title={t('pages.blocks.form.schema.description')}
              placement="left"
            >
              <InfoCircleOutlined />
            </Tooltip>
          </div>
        </div>
        <Properties
          value={field.input.value.properties}
          onChange={(v: Record<string, Schema>) =>
            field.input.onChange({
              type: 'object',
              properties: v,
            })
          }
          addLabel={t('pages.blocks.form.schema.add')}
        />
      </div>
    </div>
  );
};

const schema: Schema = {
  type: 'object',
  properties: {
    title: {
      type: 'localized:string',
      title: 'pages.blocks.settings.blockTitle.label',
      description: 'pages.blocks.settings.blockTitle.description',
    },
    schema: {
      type: 'object',
      'ui:widget': SchemaEditor,
    },
    submitLabel: {
      type: 'localized:string',
      title: 'pages.blocks.form.submitLabel.label',
    },
    onSubmit: {
      type: 'string',
      title: 'pages.blocks.form.onSubmit.label',
      description: 'pages.blocks.form.onSubmit.description',
    },
    onChange: {
      type: 'string',
      title: 'pages.blocks.form.onChange.label',
      description: 'pages.blocks.form.onChange.description',
    },
  },
};

export const Form = ({}) => {
  const { config = {}, events } = useBlock();
  const { localizeSchemaForm, localize } = useLocalizedText();
  const { t } = useTranslation('pages');

  const onChange = useCallback(
    (values: any) => {
      if (!config.onChange || !events) return;
      events.emit(config.onChange, values);
    },
    [config.onChange, events]
  );

  const onSubmit = useCallback(
    (values: any) => {
      if (!config.onSubmit || !events) return;
      events.emit(config.onSubmit, values);
    },
    [config.onSubmit, events]
  );

  const localizedSchema = useMemo(() => {
    return localizeSchemaForm(config.schema || defaultSchema);
  }, [config.schema, localizeSchemaForm]);

  return (
    <div className="p-8">
      {config.title && <BlockTitle value={config.title} />}
      <SchemaForm
        schema={localizedSchema}
        onChange={onChange}
        onSubmit={onSubmit}
        buttons={[
          <div key={0} className="flex grow justify-end mt-2 pt-4">
            <Button
              type="submit"
              variant="primary"
              className="!py-4 !px-8 !h-max"
            >
              {localize(config.submitLabel) || t('blocks.form.submit')}
            </Button>
          </div>,
        ]}
      />
    </div>
  );
};

Form.schema = schema;

export default Form;
