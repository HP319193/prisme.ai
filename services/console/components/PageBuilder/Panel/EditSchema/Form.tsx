import { FieldProps, Schema, Tooltip } from '@prisme.ai/design-system';
import { InfoCircleOutlined, PlusOutlined } from '@ant-design/icons';
import { useField } from 'react-final-form';
import { useTranslation } from 'next-i18next';
import Properties from '../../../SchemaFormBuilder/Properties';
import { Button } from 'antd';

const SchemaEditor = ({ name }: FieldProps) => {
  const field = useField(name);
  const { t } = useTranslation('workspaces');

  return (
    <div className="pr-form-field">
      <label className="pr-form-label">
        {t('pages.blocks.form.schema.label')}
      </label>
      <div className="pr-form-description">
        <Tooltip
          title={t('pages.blocks.form.schema.description')}
          placement="left"
        >
          <InfoCircleOutlined />
        </Tooltip>
      </div>
      <div className="absolute -top-2 -right-2">
        <Tooltip title={t('pages.blocks.form.schema.add')} placement="left">
          <Button
            onClick={() =>
              field.input.onChange({
                ...field.input.value,
                properties: {
                  ...field.input.value.properties,
                  '': { type: 'string' },
                },
              })
            }
          >
            <PlusOutlined />
          </Button>
        </Tooltip>
      </div>
      <div className="pr-form-input">
        <Properties
          value={field.input.value.properties}
          onChange={(v: Record<string, Schema>) =>
            field.input.onChange({
              type: 'object',
              properties: v,
            })
          }
        />
      </div>
    </div>
  );
};

const schema: Schema = {
  type: 'object',
  properties: {
    title: {
      type: 'localized:string',
      title: 'pages.blocks.settings.blockTitle.label',
      description: 'pages.blocks.settings.blockTitle.description',
    },
    schema: {
      type: 'object',
      'ui:widget': SchemaEditor,
    },
    submitLabel: {
      type: 'localized:string',
      title: 'pages.blocks.form.submitLabel.label',
      description: 'pages.blocks.form.submitLabel.description',
    },
    hideSubmit: {
      type: 'boolean',
      title: 'pages.blocks.form.hideSubmit.label',
      description: 'pages.blocks.form.hideSubmit.description',
    },
    onSubmit: {
      type: 'string',
      title: 'pages.blocks.form.onSubmit.label',
      description: 'pages.blocks.form.onSubmit.description',
    },
    onChange: {
      type: 'string',
      title: 'pages.blocks.form.onChange.label',
      description: 'pages.blocks.form.onChange.description',
    },
    disabledSubmit: {
      type: 'boolean',
      title: 'pages.blocks.form.disabledSubmit.label',
      description: 'pages.blocks.form.disabledSubmit.description',
    },
  },
};

export default schema;
