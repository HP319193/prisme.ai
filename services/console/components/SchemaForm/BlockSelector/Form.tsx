import { SchemaForm, useSchemaForm } from '@prisme.ai/design-system';
import { useBlockSelector } from './BlockSelectorProvider';

interface SchemaFormProps {
  values: Record<string, any>;
  onChange: (values: SchemaFormProps['values']) => void;
}

export const Form = ({ values, onChange }: SchemaFormProps) => {
  const { schema } = useBlockSelector();
  const { utils, locales, components } = useSchemaForm();

  if (!schema) return null;

  return (
    <SchemaForm
      schema={schema}
      locales={locales}
      buttons={[]}
      initialValues={values}
      utils={utils}
      components={components}
      onChange={onChange}
    />
  );
};

export default Form;
