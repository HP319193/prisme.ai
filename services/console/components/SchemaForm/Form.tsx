import { useTranslation } from 'next-i18next';
import { Button } from '@prisme.ai/design-system';
import { FC, ReactNode, useCallback, useMemo } from 'react';
import { Form as FFForm, FormRenderProps } from 'react-final-form';
import { Schema } from './types';
import { PlusOutlined } from '@ant-design/icons';
import arrayMutators from 'final-form-arrays';
import Layout from './Layout';

interface FormProps {
  schema: Schema;
  onSubmit: (values: any) => void;
  initialValues?: FormRenderProps['initialValues'];
  description?: string;
  submitLabel?: string | ReactNode;
  buttons?: ReactNode;
  formClassName?: string;
  formFieldsClassName?: string;
}
export const Form: FC<FormProps> = ({
  schema,
  onSubmit,
  description,
  initialValues = {},
  submitLabel,
  buttons,
  formClassName,
  formFieldsClassName,
  ...formProps
}) => {
  const { t } = useTranslation('workspaces');

  const { properties = {} } = schema;

  const required = (schema.required || []).filter((f) =>
    Object.keys(properties).includes(f)
  );

  const oneOf = schema.oneOf
    ? schema.oneOf.map(({ required = [] }) => ({
        required: required.filter((f) => Object.keys(properties).includes(f)),
      }))
    : null;

  const submit = useCallback(
    (values: any) => {
      let errors: Record<string, string> = {};
      if (
        !oneOf ||
        oneOf.some(({ required }) => required.every((f) => !!values[f]))
      ) {
        const parsedValues = Object.keys(
          schema.properties || properties
        ).reduce((prev, attr) => {
          const { type } = (schema.properties || properties)[attr];
          if (type === 'object' && values[attr]) {
            try {
              return {
                ...prev,
                [attr]: JSON.parse(values[attr]),
              };
            } catch (e) {
              return {
                ...prev,
                [attr]: values[attr],
              };
            }
          }
          if (type === 'array') {
            return {
              ...prev,
              [attr]: values[attr],
            };
          }
          return {
            ...prev,
            [attr]: values[attr],
          };
        }, {});

        if (schema.required) {
          schema.required.forEach((attr) => {
            if (!parsedValues[attr as keyof typeof parsedValues]) {
              errors[attr] = 'required';
            }
          });
        }

        if (Object.keys(errors).length === 0) {
          return onSubmit(parsedValues);
        }
      }

      // Check oneOf
      if (oneOf) {
        const oneOfIsValid = oneOf.some(({ required }) =>
          required.every((f) => values[f])
        );
        if (!oneOfIsValid) {
          oneOf.forEach(({ required }) => {
            required.forEach((f) => {
              if (values[f]) return;
              errors[f] = 'oneOfRequired';
            });
          });
        }
      }
      return errors;
    },
    [onSubmit, oneOf, properties, schema]
  );

  const fields: (Schema & { field: string })[] = useMemo(
    () =>
      Object.keys(properties).map((field) => ({
        field,
        ...properties[field],
      })),
    [properties]
  );

  return (
    <FFForm
      onSubmit={submit}
      initialValues={initialValues}
      {...formProps}
      mutators={{ ...arrayMutators }}
    >
      {({ handleSubmit }) => (
        <form onSubmit={handleSubmit} className={formClassName}>
          <div className={formFieldsClassName}>
            {description && <div>{description}</div>}
            <Layout
              options={schema['ui:options']}
              fields={fields}
              required={required}
            />
          </div>
          {buttons || (
            <Button type="submit">
              {submitLabel || (
                <>
                  <PlusOutlined />
                  {t('automations.edit.save')}
                </>
              )}
            </Button>
          )}
        </form>
      )}
    </FFForm>
  );
};

export default Form;
