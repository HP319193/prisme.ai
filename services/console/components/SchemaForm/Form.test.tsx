import Form from './Form';
import renderer from 'react-test-renderer';
import { Form as FFForm } from 'react-final-form';

it('should render', () => {
  const schema = {
    required: ['foo'],
    properties: {
      foo: {
        type: 'string',
      },
      bar: {
        type: 'boolean',
      },
      lorem: {
        type: 'object',
      },
      ipsum: {
        type: 'array',
      },
    },
  };
  const onSubmit = jest.fn();
  const root = renderer.create(<Form schema={schema} onSubmit={onSubmit} />);
  expect(root.toJSON()).toMatchSnapshot();
});

it('should submit', () => {
  const schema = {
    required: ['foo'],
    properties: {
      foo: {
        type: 'string',
      },
      bar: {
        type: 'boolean',
      },
      lorem: {
        type: 'object',
      },
      ipsum: {
        type: 'array',
      },
    },
  };
  const onSubmit = jest.fn();
  const root = renderer.create(<Form schema={schema} onSubmit={onSubmit} />);
  const errors = root.root.findByType(FFForm).props.onSubmit({});
  expect(errors).toEqual({ foo: 'required' });
});

it('should save string as string', () => {
  const schema = {
    required: ['foo'],
    properties: {
      foo: {
        type: 'object',
      },
    },
  };
  const onSubmit = jest.fn();
  const root = renderer.create(<Form schema={schema} onSubmit={onSubmit} />);
  root.root.findByType(FFForm).props.onSubmit({
    foo: 'value',
  });
  expect(onSubmit).toHaveBeenCalledWith({
    foo: 'value',
  });
});

it('should save object as object', () => {
  const schema = {
    required: ['foo'],
    properties: {
      foo: {
        type: 'object',
      },
    },
  };
  const onSubmit = jest.fn();
  const root = renderer.create(<Form schema={schema} onSubmit={onSubmit} />);
  root.root.findByType(FFForm).props.onSubmit({
    foo: '{"bar": 1}',
  });
  expect(onSubmit).toHaveBeenCalledWith({
    foo: {
      bar: 1,
    },
  });
});

it('should check oneOf rule', () => {
  const schema = {
    oneOf: [
      {
        required: ['foo', 'lorem'],
      },
      {
        required: ['bar', 'lorem'],
      },
    ],
    properties: {
      foo: {
        type: 'string',
      },
      bar: {
        type: 'string',
      },
      lorem: {
        type: 'string',
      },
    },
  };
  const onSubmit = jest.fn();
  const root = renderer.create(<Form schema={schema} onSubmit={onSubmit} />);
  const errors = root.root.findByType(FFForm).props.onSubmit({
    lorem: 'a',
  });
  expect(errors).toEqual({ foo: 'oneOfRequired', bar: 'oneOfRequired' });
});

it('should submit without error', () => {
  const schema = {
    required: ['foo'],
    properties: {
      foo: {
        type: 'string',
      },
    },
  };
  const onSubmit = jest.fn();
  const root = renderer.create(<Form schema={schema} onSubmit={onSubmit} />);
  const errors = root.root.findByType(FFForm).props.onSubmit({
    foo: 'a',
  });
  expect(onSubmit).toHaveBeenCalledWith({ foo: 'a' });
});
