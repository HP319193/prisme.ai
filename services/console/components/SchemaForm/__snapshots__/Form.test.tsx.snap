// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render 1`] = `
<form
  onSubmit={[Function]}
>
  <div
    className="mb-5 "
  >
    <span
      className="flex flex-col "
    />
  </div>
  <div
    className="mb-5"
  >
    <label
      className="mx-2 flex flex-1 align-center"
    >
      <div>
        bar
      </div>
    </label>
  </div>
  <div
    className="mb-5 "
  >
    <span
      className="flex flex-col "
    >
      <label
        className="text-gray text-[11px]"
        htmlFor="lorem"
      >
        lorem
      </label>
      <div />
    </span>
  </div>
  automations.edit.save
</form>
`;
